justfile

```diff
@@ -1,55 +1,100 @@
-    . scripts/setup_functions
-    dev_setup
+    # allow users to specify python version in .env
+    PYTHON_VERSION=${PYTHON_VERSION:-python3.9}
+    # create venv and upgrade pip
+    test -d $VIRTUAL_ENV || { $PYTHON_VERSION -m venv $VIRTUAL_ENV && $PIP install --upgrade pip; }
+    # ensure we have pip-tools so we can run pip-compile
+    test -e $BIN/pip-compile || $PIP install pip-tools
+
+
+# update requirements.prod.txt if requirement.prod.in has changed
+requirements-prod: virtualenv
+    @test requirements.prod.in -ot requirements.prod.txt || $COMPILE --output-file=requirements.prod.txt requirements.prod.in
+
+
+# update requirements.dev.txt if requirements.dev.in has changed
+requirements-dev: virtualenv
+    @test requirements.dev.in -ot requirements.dev.txt || $COMPILE --output-file=requirements.dev.txt requirements.dev.in
 
-# run the test suite. Optional args are passed to pytest
-test ARGS="":
+
+# ensure prod requirements installed and up to date
+prodenv: requirements-prod
     #!/usr/bin/env bash
-    set -euo pipefail
-    . scripts/setup_functions
-    dev_setup
+    # have the requirements changed since we last installed them?
+    test requirements.prod.txt -nt $VIRTUAL_ENV/.prod || exit 0
+    $PIP install -r requirements.prod.txt
+    touch $VIRTUAL_ENV/.prod
+    # force dev rebuild too
+    touch requirements.dev.in
 
-    python -m pytest --cov=. --cov-report html --cov-report term-missing:skip-covered {{ ARGS }}
 
-# runs the format (black), sort (isort) and lint (flake8) check but does not change any files
-check:
+# && dependencies are run after the recipe has run. Needs just>=0.9.9
+#
+# ensure dev requirements installed and up to date
+devenv: prodenv requirements-dev && install-precommit
     #!/usr/bin/env bash
-    set -euo pipefail
-    . scripts/setup_functions
-    dev_setup
+    # have the requirements changed since we last installed them?
+    test requirements.dev.txt -nt $VIRTUAL_ENV/.dev || exit 0
+    $PIP install -r requirements.dev.txt
+    touch $VIRTUAL_ENV/.dev
 
-    black --check .
-    isort --check-only --diff .
-    flake8
 
-# fix formatting and import sort ordering
-fix:
+# ensure precommit is installed
+install-precommit:
     #!/usr/bin/env bash
-    set -euo pipefail
-    . scripts/setup_functions
-    dev_setup
+    BASE_DIR=$(git rev-parse --show-toplevel)
+    test -f $BASE_DIR/.git/hooks/pre-commit || $BIN/pre-commit install
 
-    black .
-    isort .
 
-# compile and update python dependencies.  <target> specifies an environment to update (dev/prod).
-update TARGET="prod":
+# upgrade dev or prod dependencies (all by default, specify package to upgrade single package)
+upgrade env package="": virtualenv
     #!/usr/bin/env bash
-    set -euo pipefail
-    . scripts/setup_functions
-    dev_setup
+    opts="--upgrade"
+    test -z "{{ package }}" || opts="--upgrade-package {{ package }}"
+    $COMPILE $opts --output-file=requirements.{{ env }}.txt requirements.{{ env }}.in
```

Comment:
Took me a second to really grok this one, such a useful utility though!

Reply:
yeah, bash is a bit messy for this.

Reply:
But it works so nicely with the rest of this change. If you run `just upgrade dev` and then `just test`, it all just works!